File: RC_robot/RC_robot/test_RC_robot.m
%%%%%%%%%%%%%
%% 3. test %%
%%%%%%%%%%%%%

disp('testing:');

time_window = 0;

for j = 1:n_test_loops

    fprintf('  loop: %d/%d,\n', j, n_test_loops);

    % initial conditions
    Xv = 1 * (2 * rand(numUnits, 1) - 1);
    X = tanh(Xv);
    Xv_current = Xv;
    Out = zeros(numOut, 1);

    % Initialize arm state
    x0 = [zeros(2, 1); zeros(2, 1)]; %theta1, theta2, thetadot1, thetadot2
    u = zeros(numOut, 1);
    u_in = u;

    arm_x = zeros(4, n_steps);
    arm_x(:, 1) = x0 + arm_dynamics(x0, u_in, L) * dt * arm_dt;

    arm = arm_x(:, 1) + msr_test_noise_amp * randn(4, 1);
    error_cntl = 0;

    for i = 1:n_steps

        if rem(i, 100) == 0
            fprintf('    step: %d/%d\n', i, n_steps);
        end

        Input = input_pattern(:, i);

        % update RNN units
        Xv_current = W * X + WIn * Input + WFb * Out;
        Xv = Xv + ((-Xv + Xv_current) ./ tau) * dt;
        X = tanh(Xv);

        % output through readout
        Out = WOut * X + alpha * arm(1:2);
        Out_test_history(:, i, j) = Out;
        
        %PD control
        error_prev = error_cntl;
        error_cntl = Out - arm(1:2);
        dif = (error_cntl - error_prev) / arm_dt;
        u = Kp * error_cntl + Kd * dif;

        if i < start_train_n
            u = zeros(numOut, 1);
        end

        error_history(:, i, j) = error_cntl;
        u_history(:, i, j) = u;

        if strcmp(perturbation, 'impulse')
            if i == (start_train_n + 500)
                u = u - 0.5 * [1.0; 1.0];
            end
        elseif strcmp(perturbation, 'obstacle')
            if arm_x(1, i) >= 0.5 && i <= (start_train_n + 500)
                u = zeros(numOut, 1);
                arm_x(3:4, i) = zeros(numOut, 1);
            end
        end

        % System noise
        % if i >= start_train_n
        %     u = u + sys_noise_amp * randn(2, 1);
        % end

        arm_x(:, i+1) = arm_x(:, i) + arm_dynamics(arm_x(:, i), u, L) * dt * arm_dt;
        arm = arm_x(:, i+1) + msr_test_noise_amp * randn(4, 1);

        Out_test_history_link2(:, i, j) = culc_kinematics(arm_x(1:2, i), L);
        Out_test_history_link1(:, i, j) = culc_kinematics_elbow(arm_x(1:2, i), L);
    end

    % performance
    for n = 1:numOut
        R = corrcoef(Out_test_history(n, start_train_n:end_train_n, j), target_Out(n, start_train_n:end_train_n));
        R2_test(n, j) = R(1, 2)^2;
    end
    fprintf('R^2=%.3f\n', R2_test(1, j));
end

R_ave = mean(R2_test, 2);
R_std = std(R2_test');
fprintf('  mean R^2=%.3f\n', R_ave);

max_u2 = max(u_history(2, 500:1000, 1));
fprintf('  peak torque = %.5f\n', max_u2);

File: reservoir_motor/test_RC_robot.cpp
#include "test_RC_robot.h"
#include "evaluation.h"
#include "arm_dynamics.h"
#include <iostream>
#include <random>

void test_RC_robot(const Parameters& params, 
                   const Matrix& W, 
                   const Matrix& WIn, 
                   const Matrix& WFb, 
                   const Matrix& WOut,
                   const std::string& perturbation) {

    const auto& rnnParams = params.rnnParams;
    const auto& trainParams = params.trainingSettings;
    const auto& controlParams = params.robotControlParams;
    const auto& inputParams = params.inputSettings;
    const auto& outputParams = params.outputSettings;

    std::cout << "Testing:" << std::endl;

    // 乱数生成器のセットアップ
    std::default_random_engine generator(std::random_device{}());
    std::normal_distribution<double> normal_dist(0.0, 1.0);

    Matrix Out_test_history = Matrix::Zero(rnnParams.numOut, inputParams.n_steps);
    Matrix error_history = Matrix::Zero(rnnParams.numOut, inputParams.n_steps);
    Matrix u_history = Matrix::Zero(rnnParams.numOut, inputParams.n_steps);

    for (int j = 0; j < trainParams.n_test_loops; ++j) {
        std::cout << "  Loop: " << j + 1 << "/" << trainParams.n_test_loops << ", " << std::endl;

        // 初期条件の設定
        Vector Xv = Vector::Random(rnnParams.numUnits);
        Vector X = Xv.array().tanh();
        Vector Xv_current = Xv;
        Vector Out = Vector::Zero(rnnParams.numOut);

        Vector x0 = Vector::Zero(4);
        Vector u = Vector::Zero(rnnParams.numOut);
        Matrix arm_x = Matrix::Zero(4, inputParams.n_steps + 1);
        arm_x.col(0) = x0;

        Vector arm = arm_x.col(0) + trainParams.msr_test_noise_amp * Vector::Random(4);
        Vector error_cntl = Vector::Zero(rnnParams.numOut);

        for (int i = 0; i < inputParams.n_steps; ++i) {
            if (i % 100 == 0) {
                std::cout << "    Step: " << i + 1 << "/" << inputParams.n_steps << std::endl;
            }

            Vector Input = Eigen::Map<const Vector>(inputParams.input_pattern[i].data(), inputParams.input_pattern[i].size());

            // RNNの更新
            Xv_current = W * X + WIn * Input + WFb * Out;
            Xv += ((-Xv + Xv_current) / trainParams.tau) * inputParams.dt;
            X = Xv.array().tanh();

            // 出力計算
            Out = WOut * X + trainParams.alpha * arm.head(2);
            Out_test_history.col(i) = Out;

            // PD制御
            Vector error_prev = error_cntl;
            error_cntl = Out - arm.head(2);
            Vector dif = (error_cntl - error_prev) / controlParams.arm_dt;
            u = controlParams.Kp * error_cntl + controlParams.Kd * dif;

            if (i < inputParams.start_train_n) {
                u.setZero();
            }

            error_history.col(i) = error_cntl;
            u_history.col(i) = u;

            // Perturbation
            if (perturbation == "impulse" && i == (inputParams.start_train_n + 500)) {
                u -= 0.5 * Vector::Ones(rnnParams.numOut);
            } else if (perturbation == "obstacle" && arm_x(0, i) >= 0.5 && i <= (inputParams.start_train_n + 500)) {
                u.setZero();
                arm_x.block(2, i, 2, 1).setZero(); // 速度をゼロに
            }

            // システムノイズを加える
            if (i >= inputParams.start_train_n) {
                u += trainParams.sys_noise_amp * Vector::Random(rnnParams.numOut);
            }

            // アーム動力学の更新
            arm_x.col(i + 1) = arm_x.col(i) + arm_dynamics(arm_x.col(i), u, controlParams.L) * inputParams.dt;
            arm = arm_x.col(i + 1) + trainParams.msr_test_noise_amp * Vector::Random(4);
        }

        // パフォーマンス評価
        for (int n = 0; n < rnnParams.numOut; ++n) {
            double r2 = corrcoef(Out_test_history.row(n).segment(inputParams.start_train_n, inputParams.end_train_n - inputParams.start_train_n),
                                 outputParams.target_Out.row(n).segment(inputParams.start_train_n, inputParams.end_train_n - inputParams.start_train_n));
            std::cout << "R^2=" << r2 << std::endl;
        }
    }
}
